### ISSUE TITLE
Force consistent function declaration syntax

### ISSUE THREAD

**Issue by SamuelBolduc** [eslint/eslint#3838]
Given [this little test](http://jsperf.com/function-declaration-compare) I put together, wouldn't it be a good idea to have a rule to enforce declaring functions without assigning them via a `var`, `const` or `let` declaration?

In the example I linked, only the last one should be valid with this potential new rule.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
Sorry, it's pretty hard to figure out what you are proposing from that test, could you provide information requested in [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules)?

**Comment by SamuelBolduc** [comment]
Warnings:
```javascript
const fn = function(a, b) { return a + b; };
let fn = function(a, b) { return a + b; };
var fn = function(a, b) { return a + b; };
const fn = (a, b) => { return a + b; };
let fn = (a, b) => { return a + b; };
var fn = (a, b) => { return a + b; };
```
No warning:
```javascript
function fn(a, b) { return a + b; };
```

The rule doesn't really prevent an error nor is it stylistic. In some environments (Webkit for instance, so Node.js) it is much faster to declare functions via the last method (`function fn(a, b) { return a + b; };`).

On the other hand, in other environments (SpiderMonkey is an example) declaring via `const fn = (a, b) => { return a + b; };` is much faster. So it should be possible to choose which method we want to enforce.

The linked test I provided in the first post illustrates this.

**Comment by ilyavolodin** [comment]
I think what you are looking for is `func-style: [2, "expression"]` which will flag any function declaration and force function expressions instead.

**Comment by platinumazure** [comment]
@ilyavolodin Or maybe the reverse in this case, since OP seems to want declarations.

**Comment by ilyavolodin** [comment]
Right, sorry, I was looking at documentation and didn't notice that example was marked as failing. `func-style: [2, "declaration"]` is what you are looking for. Closing as there's an existing rule that covers this proposal.
